sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/BeamFnStateGrpcClientCache.java

```diff
@@ -87,9 +88,12 @@ private GrpcStateClient(ApiServiceDescriptor apiServiceDescriptor) {
       this.apiServiceDescriptor = apiServiceDescriptor;
       this.outstandingRequests = new ConcurrentHashMap<>();
       this.channel = channelFactory.apply(apiServiceDescriptor);
+      // We use the directExecutor because we just complete futures when handling responses.
+      // This showed a 1-2% improvement in the ProcessBundleBenchmark#testState* benchmarks.
       this.outboundObserver =
           outboundObserverFactory.outboundObserverFor(
-              BeamFnStateGrpc.newStub(channel)::state, new InboundObserver());
+              BeamFnStateGrpc.newStub(channel).withExecutor(MoreExecutors.directExecutor())::state,
```

Comment:
There is also the option to create a grpc channel that executes directly instead of an executor which is says can give transport ability to optimize further:
https://grpc.github.io/grpc-java/javadoc/io/grpc/ManagedChannelBuilder.html#directExecutor--

Reply:
Done.